pkg/controller/util.go

```diff
@@ -175,28 +171,79 @@ func GetVolumeCapabilities(pvSpec *v1.PersistentVolumeSpec) (*csi.VolumeCapabili
+	var err error
+	if singleNodeMultiWriterCapable {
+		am, err = getSingleNodeMultiWriterCapableAccessMode(pvSpec)
+	} else {
+		am, err = getAccessMode(pvSpec)
+	}
+	if err != nil {
+		return nil, err
+	}
+	cap.AccessMode.Mode = am
+	return cap, nil
+}
+
+// getAccessMode maps PV access modes to CSI access modes
+func getAccessMode(pvSpec *v1.PersistentVolumeSpec) (csi.VolumeCapability_AccessMode_Mode, error) {
+	m := uniqueAccessModes(pvSpec)
+
 	switch {
 	case m[v1.ReadWriteMany]:
 		// ReadWriteMany trumps everything, regardless what other modes are set
-		cap.AccessMode.Mode = csi.VolumeCapability_AccessMode_MULTI_NODE_MULTI_WRITER
+		return csi.VolumeCapability_AccessMode_MULTI_NODE_MULTI_WRITER, nil
 
 	case m[v1.ReadOnlyMany] && m[v1.ReadWriteOnce]:
-		// This is no way how to translate this to CSI...
-		return nil, fmt.Errorf("CSI does not support ReadOnlyMany and ReadWriteOnce on the same PersistentVolume")
+		// This is not possible in the CSI spec
+		return csi.VolumeCapability_AccessMode_UNKNOWN, fmt.Errorf("CSI does not support ReadOnlyMany and ReadWriteOnce on the same PersistentVolume")
 
 	case m[v1.ReadOnlyMany]:
 		// There is only ReadOnlyMany set
-		cap.AccessMode.Mode = csi.VolumeCapability_AccessMode_MULTI_NODE_READER_ONLY
+		return csi.VolumeCapability_AccessMode_MULTI_NODE_READER_ONLY, nil
 
 	case m[v1.ReadWriteOnce]:
 		// There is only ReadWriteOnce set
-		cap.AccessMode.Mode = csi.VolumeCapability_AccessMode_SINGLE_NODE_WRITER
+		return csi.VolumeCapability_AccessMode_SINGLE_NODE_WRITER, nil
 
 	default:
-		return nil, fmt.Errorf("unsupported AccessMode combination: %+v", pvSpec.AccessModes)
+		return csi.VolumeCapability_AccessMode_UNKNOWN, fmt.Errorf("unsupported AccessMode combination: %+v", pvSpec.AccessModes)
+	}
+}
+
+// getSingleNodeMultiWriterCapableAccessMode maps PV access modes to CSI access modes for drivers with the SINGLE_NODE_MULTI_WRITER
+// capability
+func getSingleNodeMultiWriterCapableAccessMode(pvSpec *v1.PersistentVolumeSpec) (csi.VolumeCapability_AccessMode_Mode, error) {
+	m := uniqueAccessModes(pvSpec)
+
+	switch {
+	case m[v1.ReadWriteMany]:
+		// ReadWriteMany trumps everything, regardless what other modes are set
+		return csi.VolumeCapability_AccessMode_MULTI_NODE_MULTI_WRITER, nil
+
+	case m[v1.ReadOnlyMany] && m[v1.ReadWriteOnce]:
+		// This is not possible in the CSI spec
+		return csi.VolumeCapability_AccessMode_UNKNOWN, fmt.Errorf("CSI does not support ReadOnlyMany and ReadWriteOnce on the same PersistentVolume")
+
+	case m[v1.ReadOnlyMany] && m[v1.ReadWriteOncePod]:
+		// This is not possible in the CSI spec
+		return csi.VolumeCapability_AccessMode_UNKNOWN, fmt.Errorf("CSI does not support ReadOnlyMany and ReadWriteOncePod on the same PersistentVolume")
+
+	case m[v1.ReadOnlyMany]:
+		// There is only ReadOnlyMany set
+		return csi.VolumeCapability_AccessMode_MULTI_NODE_READER_ONLY, nil
+
+	case m[v1.ReadWriteOnce]:
+		// There is only ReadWriteOnce set
```

Comment:
At this point,can there be both `ReadWriteOnce` and `ReadWriteOncePod` in the PV? Not sure how useful this combination is.

Reply:
Yes this is possible, but ReadWriteOncePod takes precedence since it's more restrictive. Fixed this by enforcing only one allowed to be set.